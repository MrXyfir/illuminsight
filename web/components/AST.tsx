import { Insightful } from 'types/insightful';
import * as React from 'react';

export function AST({ ast }: { ast: Insightful.AST }) {
  if (typeof ast == 'string') return <React.Fragment>{ast}</React.Fragment>;

  return React.createElement(
    ast.n,
    ast.a,
    ast.c && ast.c.map((child, i) => <AST key={i} ast={child} />)
  );
}
