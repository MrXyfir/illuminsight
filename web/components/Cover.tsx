import { createStyles, WithStyles, withStyles } from '@material-ui/core';
import { CropSquare as SquareIcon } from '@material-ui/icons';
import * as localForage from 'localforage';
import * as React from 'react';

const styles = createStyles({});

function _Cover({ classes, id }: WithStyles<typeof styles> & { id: number }) {
  const [url, setUrl] = React.useState('');

  // Load cover from local storage and create object URL from blob
  // Revoke object URL on unmount or id change
  React.useEffect(() => {
    let _url: string;
    localForage
      .getItem(`entity-cover-${id}`)
      .then(cover => {
        if (cover === null) return;
        _url = URL.createObjectURL(cover as Blob);
        setUrl(_url);
      })
      .catch(() => undefined);

    return () => {
      if (_url) URL.revokeObjectURL(_url);
    };
  }, [id]);

  return <div>{url ? <img src={url} /> : <SquareIcon />}</div>;
}

export const Cover = withStyles(styles)(_Cover);
