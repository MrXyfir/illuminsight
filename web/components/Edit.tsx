import { WithSnackbarProps, withSnackbar } from 'notistack';
import { RouteComponentProps } from 'react-router';
import * as localForage from 'localforage';
import { Insightful } from 'types/insightful';
import * as React from 'react';
import * as JSZip from 'jszip';
import {
  InputAdornment,
  createStyles,
  WithStyles,
  withStyles,
  TextField,
  Button,
  Theme
} from '@material-ui/core';
import {
  BookmarkBorder as BookmarkIcon,
  // DateRange as PublishedIcon,
  People as AuthorsIcon,
  Image as ImageIcon,
  Label as NameIcon,
  Home as PublisherIcon,
  Link as LinkIcon,
  Save as SaveIcon
} from '@material-ui/icons';

const styles = (theme: Theme) =>
  createStyles({
    coverInput: {
      display: 'none'
    },
    root: {
      padding: '1em'
    }
  });

interface EditState {
  entity?: Insightful.Entity;
  cover?: string;
}
type EditProps = WithStyles<typeof styles> &
  RouteComponentProps &
  WithSnackbarProps;

class _Edit extends React.Component<EditProps, EditState> {
  state: EditState = {};
  zip?: JSZip;

  constructor(props: EditProps) {
    super(props);
  }

  componentDidMount() {
    this.loadZip();
  }

  componentWillUnmount() {
    const { cover } = this.state;
    if (cover) URL.revokeObjectURL(cover);
  }

  /**
   * Load zip file for corresponding entity.
   */
  async loadZip() {
    const { enqueueSnackbar, history, match } = this.props;
    const { entityId } = match.params as { entityId: number };

    try {
      // Load file from localForage
      const file = await localForage.getItem(`entity-${entityId}`);
      if (file === null) throw 'Could not load data from storage';

      // Parse zip file and meta
      this.zip = await JSZip.loadAsync(file as Blob);
      const entity: Insightful.Entity = JSON.parse(
        await this.zip.file('meta.json').async('text')
      );

      // Load cover and generate blob url
      let cover;
      if (entity.cover) {
        cover = URL.createObjectURL(
          await this.zip.file(entity.cover).async('blob')
        );
      }

      // Load meta.json
      this.setState({ entity, cover });
    } catch (err) {
      // Notify user of error and send them back
      console.error(err);
      enqueueSnackbar(typeof err == 'string' ? err : 'Cannot read content');
      history.replace('/');
    }
  }

  onResetBookmark() {
    this.onChange('bookmark', {
      element: 0,
      section: 0
    } as Insightful.Entity['bookmark']);
  }

  onUploadCover(file: File) {
    const { entity, cover } = this.state;
    if (!entity || !this.zip) return;

    // Delete cover if not original
    if (entity.cover && !/^res\/\d+\./.test(entity.cover))
      this.zip.remove(entity.cover);

    // Update zip file
    entity.cover = `res/cover.${file.name.split('.').pop()}`;
    this.zip.file(entity.cover, file);
    this.zip.file('meta.json', JSON.stringify(entity));

    // Revoke old cover
    if (cover) URL.revokeObjectURL(cover);

    // Generate cover url
    this.setState({ entity, cover: URL.createObjectURL(file) });
  }

  onChange(prop: keyof Insightful.Entity, value: any) {
    if (!this.state.entity) return;
    this.setState({ entity: { ...this.state.entity, [prop]: value } });
  }

  async onSave() {
    const { entity } = this.state;
    if (!entity || !this.zip) return;

    // Update meta.json
    this.zip.file('meta.json', JSON.stringify(entity));

    // Save file
    await localForage.setItem(
      `entity-${entity.id}`,
      await this.zip.generateAsync({ type: 'blob' })
    );

    // Extract cover if available and save copy outside of zip
    if (entity.cover) {
      await localForage.setItem(
        `entity-cover-${entity.id}`,
        await this.zip.file(entity.cover).async('blob')
      );
    }

    // Save entity-list
    const entities: Insightful.Entity[] = await localForage.getItem(
      'entity-list'
    );
    const index = entities.findIndex(e => e.id == entity.id) as number;
    entities[index] = entity;
    await localForage.setItem('entity-list', entities);
  }

  render() {
    const { entity, cover } = this.state;
    const { classes } = this.props;
    if (!entity) return null;

    return (
      <form className={classes.root}>
        <TextField
          id="name"
          label="Name"
          value={entity.name}
          margin="normal"
          variant="outlined"
          onChange={e => this.onChange('name', e.target.value)}
          fullWidth
          InputProps={{
            startAdornment: (
              <InputAdornment position="start">
                <NameIcon />
              </InputAdornment>
            )
          }}
          placeholder="A Tale of Two Cities"
        />

        <TextField
          id="authors"
          label="Author(s)"
          value={entity.authors}
          margin="normal"
          variant="outlined"
          onChange={e => this.onChange('authors', e.target.value)}
          fullWidth
          InputProps={{
            startAdornment: (
              <InputAdornment position="start">
                <AuthorsIcon />
              </InputAdornment>
            )
          }}
          placeholder="Ernest Hemingway"
        />

        <TextField
          id="publisher"
          label="Publisher"
          value={entity.publisher}
          margin="normal"
          variant="outlined"
          onChange={e => this.onChange('publisher', e.target.value)}
          fullWidth
          InputProps={{
            startAdornment: (
              <InputAdornment position="start">
                <PublisherIcon />
              </InputAdornment>
            )
          }}
          placeholder="Penguin Random House"
        />

        <TextField
          id="link"
          label="Link"
          value={entity.link}
          margin="normal"
          variant="outlined"
          onChange={e => this.onChange('link', e.target.value)}
          fullWidth
          InputProps={{
            startAdornment: (
              <InputAdornment position="start">
                <LinkIcon />
              </InputAdornment>
            )
          }}
          placeholder="https://example.com/article-123"
        />

        {cover ? <img src={cover} /> : null}
        <input
          id="cover-input"
          type="file"
          multiple
          onChange={e => this.onUploadCover((e.target.files as FileList)[0])}
          className={classes.coverInput}
        />
        <label htmlFor="cover-input">
          <Button variant="text" component="span" color="secondary">
            <ImageIcon />
            Set Cover
          </Button>
        </label>

        <Button
          onClick={() => this.onResetBookmark()}
          variant="contained"
          color="primary"
        >
          <BookmarkIcon />
          Reset Bookmark
        </Button>

        <Button
          onClick={() => this.onSave()}
          variant="contained"
          color="primary"
        >
          <SaveIcon />
          Save
        </Button>
      </form>
    );
  }
}

export const Edit = withStyles(styles)(withSnackbar(_Edit));
