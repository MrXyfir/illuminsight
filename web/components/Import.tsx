import * as React from 'react';
import {
  InputAdornment,
  ListItemText,
  createStyles,
  IconButton,
  WithStyles,
  withStyles,
  TextField,
  ListItem,
  Button,
  List
} from '@material-ui/core';
import {
  InsertDriveFile as FileIcon,
  RemoveCircle as RemoveIcon,
  TextFields as TextIcon,
  Link as LinkIcon,
  Add as AddIcon
} from '@material-ui/icons';

const styles = createStyles({
  fileInput: {
    display: 'none'
  },
  fieldset: {
    marginBottom: '1.5em',
    padding: '0',
    border: 'none',
    margin: '0'
  }
});

function _Import({ classes }: WithStyles<typeof styles>) {
  const [files, setFiles] = React.useState<File[]>([]);
  const [link, setLink] = React.useState('');
  const [text, setText] = React.useState('');

  function onImportFiles() {}

  function onImportLink() {}

  function onImportText() {}

  return (
    <form onSubmit={e => e.preventDefault()}>
      <fieldset className={classes.fieldset}>
        <input
          id="file-input"
          type="file"
          multiple
          onChange={e =>
            setFiles(
              files.concat(
                [...(e.target.files as FileList)].filter(
                  f => files.findIndex(_f => f.name == _f.name) == -1
                )
              )
            )
          }
          className={classes.fileInput}
        />
        <label htmlFor="file-input">
          <Button variant="text" component="span" color="secondary">
            <FileIcon />
            Upload File
          </Button>
        </label>

        {files.length ? (
          <List dense>
            {files.map(f => (
              <ListItem key={f.name}>
                <IconButton
                  aria-label="Remove"
                  onClick={() =>
                    setFiles(files.filter(_f => _f.name != f.name))
                  }
                >
                  <RemoveIcon />
                </IconButton>
                <ListItemText primary={f.name} secondary={`${f.size} bytes`} />
              </ListItem>
            ))}
          </List>
        ) : null}

        <Button
          disabled={!files.length}
          onClick={onImportFiles}
          variant="text"
          color="primary"
        >
          <AddIcon />
          Import from Files
        </Button>
      </fieldset>

      <fieldset className={classes.fieldset}>
        <TextField
          id="link"
          label="Link"
          value={link}
          margin="normal"
          variant="outlined"
          onChange={e => setLink(e.target.value)}
          fullWidth
          onKeyDown={e => (e.key == 'Enter' ? onImportLink() : null)}
          helperText="Import article content from a web page"
          InputProps={{
            startAdornment: (
              <InputAdornment position="start">
                <LinkIcon />
              </InputAdornment>
            )
          }}
          placeholder="https://example.com/article-123"
        />
        <Button
          disabled={!link}
          onClick={onImportLink}
          variant="text"
          color="primary"
        >
          <AddIcon />
          Import from Link
        </Button>
      </fieldset>

      <fieldset className={classes.fieldset}>
        <TextField
          id="text"
          label="Text"
          value={text}
          margin="normal"
          rowsMax={5}
          variant="outlined"
          onChange={e => setText(e.target.value)}
          fullWidth
          multiline
          onKeyDown={e => (e.key == 'Enter' ? onImportText() : null)}
          helperText="Paste text content"
          InputProps={{
            startAdornment: (
              <InputAdornment position="start">
                <TextIcon />
              </InputAdornment>
            )
          }}
          placeholder="Paste content here..."
        />
        <Button
          disabled={!text}
          onClick={onImportText}
          variant="text"
          color="primary"
        >
          <AddIcon />
          Import from Text
        </Button>
      </fieldset>
    </form>
  );
}

export const Import = withStyles(styles)(_Import);
