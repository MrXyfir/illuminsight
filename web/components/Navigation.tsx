import { withSnackbar, withSnackbarProps } from 'notistack';
import * as React from 'react';
import { Link } from 'react-router-dom';
import {
  createStyles,
  WithStyles,
  withStyles,
  IconButton,
  Typography,
  Tooltip,
  Toolbar,
  AppBar,
  Theme
} from '@material-ui/core';
import {
  LibraryBooks as LibraryIcon,
  LibraryAdd as ImportIcon,
  Brightness2 as MoonIcon,
  WbSunny as SunIcon
} from '@material-ui/icons';

const styles = (theme: Theme) =>
  createStyles({
    title: {
      flexGrow: 1
    }
  });

class _Navigation extends React.Component<
  WithStyles<typeof styles> & withSnackbarProps
> {
  onTheme(dark: boolean) {
    localStorage.theme = dark ? 'dark' : 'light';
    location.reload();
  }

  render() {
    const { classes } = this.props;
    return (
      <AppBar position="fixed" color="default">
        <Toolbar>
          <Typography
            className={classes.title}
            variant="h6"
            noWrap
            color="inherit"
          >
            Insightful
          </Typography>
          <Tooltip title="Toggle light/dark theme">
            <IconButton
              onClick={() => this.onTheme(localStorage.theme != 'dark')}
            >
              {localStorage.theme == 'dark' ? <SunIcon /> : <MoonIcon />}
            </IconButton>
          </Tooltip>
          <Tooltip title="Import content into library">
            <Link to="/import">
              <IconButton>
                <ImportIcon />
              </IconButton>
            </Link>
          </Tooltip>
          <Tooltip title="Search items in library">
            <Link to="/library">
              <IconButton>
                <LibraryIcon />
              </IconButton>
            </Link>
          </Tooltip>
        </Toolbar>
      </AppBar>
    );
  }
}

export const Navigation = withSnackbar(withStyles(styles)(_Navigation));
