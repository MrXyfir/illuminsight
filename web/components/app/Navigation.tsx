import { Route, Link } from 'react-router-dom';
import * as React from 'react';
import {
  createStyles,
  WithStyles,
  withStyles,
  IconButton,
  Typography,
  Tooltip,
  Toolbar,
  AppBar
} from '@material-ui/core';
import {
  LibraryBooks as LibraryIcon,
  LibraryAdd as ImportIcon,
  Brightness2 as MoonIcon,
  WbSunny as SunIcon,
  Edit as EditIcon
} from '@material-ui/icons';

const styles = createStyles({ title: { flexGrow: 1 } });

class _Navigation extends React.Component<WithStyles<typeof styles>> {
  onTheme(dark: boolean) {
    localStorage.theme = dark ? 'dark' : 'light';
    location.reload();
  }

  render() {
    const { classes } = this.props;
    return (
      <AppBar position="fixed" color="default">
        <Toolbar>
          <Typography
            className={classes.title}
            variant="h6"
            noWrap
            color="inherit"
          >
            Insightful
          </Typography>

          <Tooltip title="Toggle light/dark theme">
            <IconButton
              onClick={() => this.onTheme(localStorage.theme != 'dark')}
            >
              {localStorage.theme == 'dark' ? <SunIcon /> : <MoonIcon />}
            </IconButton>
          </Tooltip>
          <Tooltip title="Import content into library">
            <Link to="/import">
              <IconButton>
                <ImportIcon />
              </IconButton>
            </Link>
          </Tooltip>
          <Tooltip title="Search items in library">
            <Link to="/library">
              <IconButton>
                <LibraryIcon />
              </IconButton>
            </Link>
          </Tooltip>
          <Route
            path="/read/:pub"
            render={props => (
              <Tooltip title="Edit metadata">
                <Link to={`/edit/${props.match.params.pub}`}>
                  <IconButton>
                    <EditIcon />
                  </IconButton>
                </Link>
              </Tooltip>
            )}
          />
        </Toolbar>
      </AppBar>
    );
  }
}

export const Navigation = withStyles(styles)(_Navigation);
