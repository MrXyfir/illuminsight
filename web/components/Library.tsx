import { Search as SearchIcon, Star as StarIcon } from '@material-ui/icons';
import { formatRelative } from 'date-fns';
import * as localForage from 'localforage';
import { Insightful } from 'types/insightful';
import * as React from 'react';
import { Cover } from 'components/Cover';
import * as Fuse from 'fuse.js';
import { Link } from 'react-router-dom';
import {
  ListItemAvatar,
  InputAdornment,
  createStyles,
  ListItemText,
  WithStyles,
  Typography,
  withStyles,
  TextField,
  ListItem,
  Avatar,
  Theme,
  Chip,
  List
} from '@material-ui/core';

const styles = (theme: Theme) =>
  createStyles({
    entityName: {
      fontSize: '150%',
      display: 'flex'
    },
    entityInfo: {
      color: theme.palette.grey[500]
    },
    entity: {
      textDecoration: 'none'
    }
  });

function _Library({ classes }: WithStyles<typeof styles>) {
  const [selectedTags, setSelectedTags] = React.useState<
    Insightful.Tag['id'][]
  >([]);
  const [entities, setEntities] = React.useState<Insightful.Entity[]>([]);
  const [search, setSearch] = React.useState('');
  const [tags, setTags] = React.useState<Insightful.Tag[]>([]);
  const now = new Date();

  // Load entities and tags from local storage on mount
  React.useEffect(() => {
    localForage
      .getItem('tag-list')
      .then(tags => {
        if (tags !== null) setTags(tags as Insightful.Tag[]);
        return localForage.getItem('entity-list');
      })
      .then(entities => setEntities((entities as Insightful.Entity[]) || []))
      .catch(() => undefined);
  }, []);

  // Filter by tags
  let matches = selectedTags.length
    ? entities.filter(
        entity => entity.tags.findIndex(tag => selectedTags.includes(tag)) > -1
      )
    : entities;

  // Filter by search
  if (search) {
    const fuse = new Fuse(matches, {
      shouldSort: true,
      threshold: 0.4,
      keys: ['name', 'link', 'authors', 'published', 'publisher']
    });
    matches = fuse.search(search);
  }

  // Starred items stick to top
  matches = matches.sort((a, b) => {
    if (!a.starred && b.starred) return 1;
    if (a.starred && !b.starred) return -1;
    return 0;
  });

  return (
    <div>
      {/* Search field */}
      <TextField
        id="search"
        label="Search"
        value={search}
        margin="normal"
        variant="outlined"
        onChange={e => setSearch(e.target.value)}
        fullWidth
        InputProps={{
          startAdornment: (
            <InputAdornment position="start">
              <SearchIcon />
            </InputAdornment>
          )
        }}
        placeholder="A Tale of Two Cities"
      />

      {/* Display selected tags */}
      {selectedTags.map(tag => (
        <Chip
          key={tag}
          label={(tags.find(t => t.id == tag) as Insightful.Tag).name}
          avatar={<Avatar>#</Avatar>}
          variant="outlined"
          onDelete={() => setSelectedTags(selectedTags.filter(t => t != tag))}
        />
      ))}

      {/* Display unselected tags linked to matching entities */}
      <List dense>
        {Array.from(
          new Set(
            matches
              .map(m => m.tags)
              .flat()
              .filter(t => !selectedTags.includes(t))
          )
        ).map(tag => (
          <ListItem
            key={tag}
            button
            onClick={() => setSelectedTags(selectedTags.concat(tag))}
          >
            <ListItemAvatar>
              <Avatar>#</Avatar>
            </ListItemAvatar>
            <ListItemText
              primary={(tags.find(t => t.id == tag) as Insightful.Tag).name}
            />
          </ListItem>
        ))}
      </List>

      {/* Display matching entities */}
      <List dense>
        {matches.map(match => (
          <Link to={`/read/${match.id}`} className={classes.entity}>
            <ListItem key={match.id} button>
              <Cover id={match.id} />
              <div>
                <Typography className={classes.entityName} variant="h2">
                  {match.starred ? <StarIcon color="primary" /> : null}
                  {match.name}
                </Typography>
                <Typography className={classes.entityInfo}>
                  {match.words} words â€” Added {formatRelative(match.id, now)}
                </Typography>
                <Typography className={classes.entityInfo}>
                  {match.tags
                    .map(
                      tag =>
                        `#${
                          (tags.find(t => t.id == tag) as Insightful.Tag).name
                        }`
                    )
                    .join(' ')}
                </Typography>
              </div>
            </ListItem>
          </Link>
        ))}
      </List>
    </div>
  );
}

export const Library = withStyles(styles)(_Library);
