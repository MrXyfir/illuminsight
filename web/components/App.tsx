import { BrowserRouter, Switch, Route } from 'react-router-dom';
import { SnackbarProvider } from 'notistack';
import { Navigation } from 'components/Navigation';
import { Library } from 'components/Library';
import { Import } from 'components/Import';
import { Reader } from 'components/Reader';
import * as React from 'react';
import { theme } from 'lib/theme';
import {
  MuiThemeProvider,
  createStyles,
  CssBaseline,
  WithStyles,
  withStyles,
  Button,
  Theme
} from '@material-ui/core';

const styles = (theme: Theme) =>
  createStyles({
    toolbar: theme.mixins.toolbar,
    main: {
      flexDirection: 'column',
      overflow: 'auto',
      display: 'flex',
      height: '100vh'
    }
  });

const _App = ({ classes }: WithStyles<typeof styles>) => (
  <MuiThemeProvider theme={theme}>
    <CssBaseline />
    <SnackbarProvider
      action={[
        <Button color="primary" size="small">
          Dismiss
        </Button>
      ]}
      maxSnack={2}
      anchorOrigin={{ vertical: 'top', horizontal: 'center' }}
    >
      <BrowserRouter>
        <div>
          <Navigation />
          <main className={classes.main}>
            <div className={classes.toolbar} />
            <Switch>
              <Route path="/read/:entityId" component={Reader} />
              <Route path="/library" component={Library} />
              <Route path="/import" component={Import} />
            </Switch>
          </main>
        </div>
      </BrowserRouter>
    </SnackbarProvider>
  </MuiThemeProvider>
);

export const App = withStyles(styles)(_App);
