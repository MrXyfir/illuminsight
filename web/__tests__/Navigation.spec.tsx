import { StaticRouter } from 'react-router-dom';
import { Navigation } from 'components/Navigation';
import { create } from 'react-test-renderer';
import * as React from 'react';

test('<Navigation>', () => {
  expect(
    create(
      <StaticRouter>
        <Navigation />
      </StaticRouter>
    ).toJSON()
  ).toMatchSnapshot();
});
