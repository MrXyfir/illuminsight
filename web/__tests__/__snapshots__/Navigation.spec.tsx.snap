// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation> 1`] = `
<header
  className="MuiPaper-root-11 MuiPaper-elevation4-17 MuiAppBar-root-2 MuiAppBar-positionFixed-3 MuiAppBar-colorDefault-8 mui-fixed"
>
  <div
    className="MuiToolbar-root-38 MuiToolbar-regular-40 MuiToolbar-gutters-39"
  >
    <h6
      className="MuiTypography-root-42 MuiTypography-h6-59 MuiTypography-colorInherit-71 MuiTypography-noWrap-68 _Navigation-title-1"
    >
      Insightful
    </h6>
    <button
      aria-describedby={null}
      className="MuiButtonBase-root-92 MuiIconButton-root-86"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      title="Toggle light/dark theme"
      type="button"
    >
      <span
        className="MuiIconButton-label-91"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-95"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 2c-1.82 0-3.53.5-5 1.35C7.99 5.08 10 8.3 10 12s-2.01 6.92-5 8.65C6.47 21.5 8.18 22 10 22c5.52 0 10-4.48 10-10S15.52 2 10 2z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root-104"
      />
    </button>
    <a
      aria-describedby={null}
      className=""
      href="/import"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      title="Import content into library"
    >
      <button
        className="MuiButtonBase-root-92 MuiIconButton-root-86"
        disabled={false}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-91"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-95"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M4 6H2v14c0 1.1.9 2 2 2h14v-2H4V6zm16-4H8c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-1 9h-4v4h-2v-4H9V9h4V5h2v4h4v2z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-104"
        />
      </button>
    </a>
    <a
      aria-describedby={null}
      className=""
      href="/library"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      title="Search items in library"
    >
      <button
        className="MuiButtonBase-root-92 MuiIconButton-root-86"
        disabled={false}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-91"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-95"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M4 6H2v14c0 1.1.9 2 2 2h14v-2H4V6zm16-4H8c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-1 9H9V9h10v2zm-4 4H9v-2h6v2zm4-8H9V5h10v2z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-104"
        />
      </button>
    </a>
  </div>
</header>
`;
